Terms 2.0: son of Terms (#71)
JuliaStats/StatsModels.jl: docs/src/formula.md

`StatsModels` is to make it simpler to transform tabular data into matrix format
suitable for statistical modeling.

At the moment, "tabular data" means a
[Tables.jl](https://github.com/JuliaData/Tables.jl) table, which will be
materialized as a `Tables.ColumnTable` (a `NamedTuple` of column vectors).  Work
on first-class support for streaming/row-oriented tables is ongoing.

## The `@formula` DSL

StatsModels implements the [`@formula`](@ref) domain-specific language (DSL) for
describing table-to-matrix transformations.  This DSL is designed to be familiar
to users of other statistical software, while also taking advantage of Julia's
unique strengths to be fast and flexible.

> A basic formula is composed of individual *terms*—symbols which refer to data

Reviewer: Add spaces around dashes here and below? I first read this as "terms-symbols". Or maybe change the syntax.

Author: hard to tell with the monospaced font but those are em dashes so they _should_ be sufficiently visually distinct (unless of course you've built the docs and looked there, in which case I'm willing to reconsider :))

Reviewer: I guess that's a difference between English and French typography rules...

columns, or literal numbers `0` or `1`—combined by `+`, `&`, `*`, and (at the
top level) `~`.

!!! note 

    The `@formula` macro **must** be called with parentheses to ensure that
    the formula is parsed properly.

> ### An example

Reviewer: Since that's the only heading at that level, I'd turn it into a normal sentence, e.g. "We now illustrate this with an example."

```@repl 1
using Random; Random.seed!(1); # hide
using StatsModels, DataFrames
f = @formula(y ~ 1 + a + b + c + b&c)

> df = DataFrame(y = rand(9), a = [1:9;], b = rand(9), c = repeat(["d","e","f"], 3))

Reviewer: ```suggestion
df = DataFrame(y = rand(9), a = 1:9, b = rand(9), c = repeat(["d","e","f"], 3))
```

f = apply_schema(f, schema(f, df))
response, predictors = model_cols(f, df); 
predictors
coefnames(f)
```

Let's break down the formula expression ` y ~ 1 + a + b + c + b&c`:

At the top level is the **formula separator** `~`, which separates the left-hand

> (or response) variables `y` from the right-hand size (or predictor) variables on

Reviewer: ```suggestion
(or response) variable `y` from the right-hand size (or predictor) variables on
```

the right `1 + a + b + c + b&c`.

The left-hand side has one term `y` which means that the resopnse variable is
the column from the data named `:y`.  The response can be accessed with the
analogous `response(f, df)` function.

The right hand side is made up of a number of different **terms**, separated by
`+`: `1 + a + b + c + b&c`.  Each term corresponds to one or more columns in the
generated model matrix: 

* The first term `1` generates a constant or "intercept" column full of `1.0`s.
* The next two terms `a` and `b` correspond to columns from the data table
  called `:a`, `:b`, which both hold numeric data (`Float64` and `Int`
  respectively).  Such **continuous terms** are converted to `Float64` and

>   copied to the model matrix

Reviewer: ```suggestion
  copied to the model matrix.
```

* The term `c` corresponds to the `:c` column in the table, which is _not_
  numeric, so it has been [contrast coded](@ref Modeling-categorical-data):
  there are three unique values or levels, and the default coding scheme
  ([`DummyCoding`](@ref)) generates an indicator variable for each level after

>   the first (e.g., `df[:c] .== "b"` and `df[:c] .== "a"`).

Reviewer: ```suggestion
  the first, e.g. `df[:c] .== "b"` and `df[:c] .== "a"`).
```

Author: did you mean to delete that paren there?

Reviewer: Yes, there's one opening parenthesis without a closing one in the sentence.

* The last term `b&c` is an **interaction term**, and generates model matrix
  columns for each _pair_ of columns generated by the `b` and `c` terms.

>   Columns are combined with elementwise multiplication.  Since `b` generates

Reviewer: ```suggestion
  Columns are combined with element-wise multiplication.  Since `b` generates
```

Also elsewhere.

  only a single column and `c` two, `b&c` generates two columns, equivalent to
  `df[:b] .* (df[:c] .== "b")` and `df[:b] .* (df[:c] .== "c")`.

Because we often want to include both "main effects" (`b` and `c`) and
interactions (`b&c`) of multiple variables, the formula DSL uses the `*`
operator to denote this:

```@repl 1
@formula(y ~ 1 + a + b*c)
```

Also note that the interaction operators `&` and `*` are distributive with the
term separator `+`:

```@repl 1
@formula(y ~ 1 + (a + b) & c)
```

## Non-DSL calls

Any calls to functions that are not built into the DSL (or part of an
[extension](@ref Internals-and-extending-the-formula-DSL) provided by a modeling
package) are treated like normal Julia code, and evaluated elementwise:

```@repl 1

> modelmatrix(@formula(y ~ 1 + a + log(1+a) + length(c)), df)

Reviewer: What is `length(c)` translated to in the matrix? It's a bit confusing for me since it will give a single constant, but the formula only allows `+1`.

Author: I was being a little cheeky.  It will compute the length of each string in the `:c` column and insert those values in a column (which in this case of course is perfectly collinear with the intercept...)

Reviewer: Ah, of course. Maybe take another example which is more typical of common uses?

Author: I've now expanded this example with a separate example of a function you might (perhaps) want to actually use for strings, and explicitly flagged it as such...

```

Note that the expression `1 + a` is treated differently as part of the formula
than in the call to `log`, where it's interpreted as normal addition.

## Constructing a formula programatically

A formula can be constructed at run-time by creating `Term`s and combining them
with the formula operators `+`, `&`, and `~`:

```@repl 1
Term(:y) ~ ConstantTerm(1) + Term(:a) + Term(:b) + Term(:a) & Term(:b)
```

The [`term`](@ref) function constructs a term of the appropriate type from
symbols and numbers, which makes it easy to work with collections of mixed type:

```@repl 1
ts = term(1, :a, :b)
```

> These can then be combine with standard reduction techniques:

Reviewer: ```suggestion
These can then be combined with standard reduction techniques:
```

```@repl 1
f1 = term(:y) ~ foldl(+, ts)
f2 = term(:y) ~ sum(ts)
f1 == f2 == @formula(y ~ 1 + a + b)
```

## Fitting a model from a formula

The main use of `@formula` is to streamline specifying and fitting statistical
models based on tabular data.  From the user's perspective, this is done by
`fit` methods that take a `FormulaTerm` and a table instead of numeric
matrices.

As an example, we'll simulate some data from a linear regression model with an
interaction term, a continuous predictor, a categorical predictor, and the
interaction of the two, and then fit a `GLM.LinearModel` to recover the
simulated coefficients.

```julia-repl
julia> using GLM, DataFrames, StatsModels

julia> data = DataFrame(a = rand(100), b = repeat(["d", "e", "f", "g"], 25));

# simulate response:
julia> X = modelmatrix(@formula(y ~ 1 + a*b).rhs, data);

> julia> β_true = [1., 2, 3, 4, 5, 6, 7, 8];

Reviewer: ```suggestion
julia> β_true = 1:8;
```

Author: I never know where I can safely use a range in place of a vector so I just superstitiously always use a vector (as you've seen)

Reviewer: These days ranges should work in most situations AFAIK. The main difference used to be equality, but we've fixed that.

julia> ϵ = randn(100)*0.1;

julia> data[:y] = X*β_true .+ ϵ

julia> fit(LinearModel, @formula(y ~ 1 + a*b), data)
StatsModels.TableRegressionModel{LinearModel{LmResp{Array{Float64,1}},DensePredChol{Float64,LinearAlgebra.Cholesky{Float64,Array{Float64,2}}}},Array{Float64,2}}

> y (continuous) ~ 1 + a (continuous) + b (4 levels): DummyCoding(3) + a (continuous) & b (4 levels): DummyCoding(3)

Reviewer: Isn't this too verbose? Even for a simple model like here, the line doesn't fit on a standard screen. Also in this context, we can just check the coefficient table below to see how each variable has been expanded.

Author: yeah I've considered that.  one option would be to add a check for a compact io context (and use that when showing a formula term or when showing a model) and just print something abbreviated.

Author: (and by "considered" I mean also worried about it, but haven't acted in those worries yet...)

Reviewer: Good idea.

Though in addition to that the verbose printing could be compressed a bit. Continuous variables could be treated as the default, so that they don't need a specific mention (and indeed they correspond to a single column in the matrix). And for categorical variables, the number of levels doesn't have to be repeated in addition to the coding (even though it can be kept when printing the term in isolation). Finally, to make it easier to parse the formula, it would make sense to have spaces only around operators, but not between a term and its description.

a             2.04038 0.0781928 26.0942   <1e-43
b: e          3.06689 0.0585607 52.3711   <1e-69
b: f          4.03094 0.0620725 64.9393   <1e-77
b: g          5.06553 0.0656504 77.1592   <1e-84
a & b: e      5.90041  0.101969 57.8648   <1e-73
a & b: f       6.9842  0.105653 66.1048   <1e-78
a & b: g      7.93213  0.108797 72.9077   <1e-82
```

Internally, this is accomplished in three steps:

1. The expression passed to `@formula` is lowered to term constructors combined
   by `~`, `+`, and `&`, which evaluate to create terms for the whole formula
   and any interaction terms.
2. A schema is extracted from the data, which determines whether each variable
   is continuous or categorical and extracts the summary statistics of each
   variable (mean/variance/min/max or unique levels respectively).  This schema
   is then _applied_ to the formula with `apply_schema(term, schema,
   ::Type{Model})`, which returns a new formula with each placeholder `Term`
   replaced with a concrete `ContinuousTerm` or `CategoricalTerm` as
   appropriate.  This is also the stage where any custom syntax is applied (see
   [the section on extending the `@formula`

>    DSL](@ref Internals-and-extending-the-formula-DSL) for more details)

Reviewer: ```suggestion
   DSL](@ref Internals-and-extending-the-formula-DSL) for more details).
```

3. Numeric arrays are generated for the response and predictors from the full

>    table using `model_cols(term, data)`.

Reviewer: Question about the terminology: isn't it slightly inconsistent to use an imperative verb for `apply_schema`, but only nouns for `model_cols`?

Author: perhaps but I couldn't think of something imperative and concise.  also there's teh convention that a noun named function means "make one of these from your arguments" so that's kind of what I was going for.  same with `schema`.  Do you have another suggestion?

Reviewer: Naming is hard. After thinking more about it, I think one of the reasons I find the names inconsistent is that `apply_schema` describes the input (schema), but not the output (contrary to `model_cols`, as you note). Can we find a name describing the result of `apply_schema`? `model_terms`?

Author: my original idea was `instantiate` but I thought that was maybe too vague

`StatsModels` is to make it simpler to transform tabular data into matrix format
suitable for statistical modeling.

At the moment, "tabular data" means a
[Tables.jl](https://github.com/JuliaData/Tables.jl) table, which will be
materialized as a `Tables.ColumnTable` (a `NamedTuple` of column vectors).  Work
on first-class support for streaming/row-oriented tables is ongoing.

## The `@formula` DSL

StatsModels implements the [`@formula`](@ref) domain-specific language (DSL) for
describing table-to-matrix transformations.  This DSL is designed to be familiar
to users of other statistical software, while also taking advantage of Julia's
unique strengths to be fast and flexible.

A basic formula is composed of individual *terms*—symbols which refer to data
columns, or literal numbers `0` or `1`—combined by `+`, `&`, `*`, and (at the
top level) `~`.

!!! note 

    The `@formula` macro **must** be called with parentheses to ensure that
    the formula is parsed properly.

Here is an example of the `@formula` in action:

> ```@repl 1

Reviewer: It would be nice to turn these into `jldoctest` blocks so that any change in output it detected. Can be done later though.

Author: shouldn't be too hard.  I think there's a documenter setting that will actually generate the doctest format, or I can just copy and paste from the markdown output

Author: okay, all doctests now!  the added bonus of that is that the actual output is included in the source files so it's a net win for sure

[extension](@ref Internals-and-extending-the-formula-DSL) provided by a modeling
package) are treated like normal Julia code, and evaluated elementwise:

```@repl 1
modelmatrix(@formula(y ~ 1 + a + log(1+a)), df)
```

Note that the expression `1 + a` is treated differently as part of the formula
than in the call to `log`, where it's interpreted as normal addition.

This even applies to custom functions.  For instance, if for some reason you
wanted to include a regressor based on a `String` column that encoded whether
any character in a string was after `'e'` in the alphabet, you could do

```@repl 1
gt_e(s) = any(c > 'e' for c in s)
modelmatrix(@formula(y ~ 1 + gt_e(c)), df)
```

Julia functions like this are evaluated elementwise when the numeric arrays are
created for the response and model matrix.  This makes it easy to fit models to
transformed data _lazily_, without creating temporary columns in your table.
For instance, to fit a linear regression to a log-transformed response:

```@repl 1
using GLM
lm(@formula(log(y) ~ 1 + a + b), df)
df[:log_y] = log.(df[:y]);
lm(@formula(log_y ~ 1 + a + b), df)            # equivalent
```

> The no-op function `identity` can be used to block the formula DSL

Reviewer: Damn, DSL reappeared here! :-)

Author: oh rats!

 1.0  6.0  0.28119    0.0  1.0  0.0       0.28119
 1.0  7.0  0.209472   0.0  0.0  0.0       0.0
 1.0  8.0  0.251379   1.0  0.0  0.251379  0.0
 1.0  9.0  0.0203749  0.0  1.0  0.0       0.0203749

julia> coefnames(f)
("y", ["(Intercept)", "a", "b", "c: e", "c: f", "b & c: e", "b & c: f"])

```

Let's break down the formula expression ` y ~ 1 + a + b + c + b&c`:

At the top level is the **formula separator** `~`, which separates the left-hand
(or response) variable `y` from the right-hand size (or predictor) variables on
the right `1 + a + b + c + b&c`.

The left-hand side has one term `y` which means that the resopnse variable is
the column from the data named `:y`.  The response can be accessed with the
analogous `response(f, df)` function.

The right hand side is made up of a number of different **terms**, separated by
`+`: `1 + a + b + c + b&c`.  Each term corresponds to one or more columns in the
generated model matrix: 

* The first term `1` generates a constant or "intercept" column full of `1.0`s.
* The next two terms `a` and `b` correspond to columns from the data table
  called `:a`, `:b`, which both hold numeric data (`Float64` and `Int`

>   respectively).  Such **continuous terms** are converted to `Float64` and

Reviewer 2: **numerical**?

Author: instead of "continuous terms"? or instead of numeric?

Reviewer 2: Instead of "continuous"

Author: A numerical term isn't necessarily continuous.  For instance, you might want to treat a vector Ints as indices instead of numbers.

Author: But this distinction could be made clearer here I think...

> variables.  The `*` operator does this, expanding in the following way:

Reviewer 3: Can some more words be added here to explain what `*` is ?

Reviewer 3: It isn't even entirely clear to me (as someone with out an R background) what `&` is saying

Reviewer 3: the example kind of makes it clear what `*` does, but at least there must be some words to describe it.
It probanly should be a seperate paragraph to `&`
